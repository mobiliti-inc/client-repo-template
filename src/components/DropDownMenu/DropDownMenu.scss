@import "~variables";

.dropdown-menu {
    display: block;
    width: 100%;
    display: grid;
}

.dropdown-menu-tile {
    margin-bottom: 11px;
    display: block;
    .dropdown-menu-modal-open, .dropdown-menu-modal-close {
        box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
        &:hover {
            box-shadow: 1px 2px 15px rgba(0, 0, 0, 0.32), 0 1px 2px rgba(0, 0, 0, 0.24);
        }
    }
    .dropdown-menu-modal-close {
        border-radius: 4px;
    }
    .dropdown-menu-modal-open {
        border-radius: 4px 4px 0px 0px;
    }
    .dropdown-menu-content {
        border-radius: 0px 0px 4px 4px;
    }
}

.dropdown-menu-modal {
    display: grid;
    grid-template-columns: 3fr 1fr;
    grid-column-gap: 0.5rem;
    align-items: center;
    cursor: pointer;
    height: 100px;
    width: 100%;
    background-color: $white;
    border-radius: 2px;
    padding: 0 10%;

    @media (max-width: $half-mobile-width) {
        grid-template-columns: 8fr 1fr;
    }
}

.dropdown-menu-content {
    visibility: hidden;
    background-color: $white;
    width: 100%;
    box-shadow: 0px 3px 10px lightgrey;
    z-index: 1;
    transition: all 0.5s ease-in-out;
    -webkit-transition: all 0.5s ease-in-out;
    border-radius: 2px;
    height: 100%;
    margin-bottom: 11px;
    overflow-y: scroll;
}

.dropdown-menu-modal-open {
    border-bottom: 1px rgba(0, 0, 0, 0.1) solid;
    box-shadow: 0px 0px 10px lightgrey;

    &+.dropdown-menu-content {
        visibility: visible;
        opacity: 1;
    }
}

.dropdown-menu-modal-close {
    box-shadow: 0px 3px 10px lightgrey;

    &+.dropdown-menu-content {
        visibility: hidden;
        opacity: 0;
        height: 0;
    }
}

.left-content-container {
    display: block;
    color: $blue;

    &-split {
        color: $blue;
        display: flex;
        align-items: center;
    }

    &-split>h2 {
        border-right: 1px rgba(0, 0, 0, 0.1) solid;
        padding-right: 5%;
    }

    &-split>p {
        padding-left: 5%
    }
}

.left-content-container-selected {
    color: $darker-blue;
}

.right-content-container {
    display: block;
    justify-self: end;
}

.dropdown-menu-title {
    color: inherit;
    font-family: $hanken-sans;
    font-size: 1rem;
    font-weight: bold;
    letter-spacing: -0.62px;
    margin: 0;

    @media (max-width: $mobile-width) {
        font-size: 0.9rem;
    }
}

.dropdown-menu-description {
    opacity: 0.8;
    color: inherit;
    font-family: $hanken-sans;
    font-size: 1rem;
    margin: 0;

    @media (max-width: $mobile-width) {
        font-size: 0.7rem;
    }
}