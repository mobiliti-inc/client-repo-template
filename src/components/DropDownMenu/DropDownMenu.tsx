import classNames from 'classnames';
import styles from './DropDownMenu.scss';
import { chevronUpDark, chevronDownGrey } from '../../assets/icons';

type DropDownMenuProps = {
	leftContentIsSplitted?: boolean,
	isContentCompleted?: boolean,
	menuTitle?: React.ReactNode | string,
	menuDescription?: React.ReactNode | string,
	isDroppedDown?: boolean,
	tile?: boolean,
	bindIsDroppedDownStateToParent?: (...args: any[]) => any
};

type DropDownMenuState = {
	isDroppedDown: any,
	isContentCompleted: any
};
class DropDownMenu extends React.Component<DropDownMenuProps, DropDownMenuState> {
	constructor(props) {
		super(props);
		this.state = {
			isDroppedDown: props.isDroppedDown,
			isContentCompleted: props.isContentCompleted
		};
	}

	componentDidMount = () => {
		this.props.bindIsDroppedDownStateToParent({ isDroppedDown: this.state.isDroppedDown });
	};

	componentDidUpdate = (prevProps, prevState) => {
		const { isDroppedDown, isContentCompleted } = this.props;
		if (prevProps.isDroppedDown !== isDroppedDown) {
			this.setState({
				isDroppedDown
			});
		}
		if (prevState.isContentCompleted !== isContentCompleted) {
			this.setState({
				isContentCompleted
			});
		}
	};

	handleClick = () => {
		this.setState(currentState => {
			const { isDroppedDown } = currentState;
			this.props.bindIsDroppedDownStateToParent({ isDroppedDown: !isDroppedDown });
			return {
				isDroppedDown: !isDroppedDown
			};
		});
	};

	handleDropDownClass = () => {
		const { isDroppedDown } = this.state;
		return isDroppedDown ? 'dropdown-menu-modal-open' : 'dropdown-menu-modal-close';
	};

	handleLeftContentcontainerClass = () => {
		const { leftContentIsSplitted } = this.props;
		const { isContentCompleted, isDroppedDown } = this.state;
		let leftContentcontainerClass = leftContentIsSplitted ? 'left-content-container-split' : 'left-content-container';
		leftContentcontainerClass += isContentCompleted || isDroppedDown ? ' left-content-container-selected' : '';
		return leftContentcontainerClass;
	};

	renderDropDownIcon = () => {
		const { isDroppedDown } = this.state;
		return <img src={isDroppedDown ? chevronUpDark : chevronDownGrey} alt="drop icon" />;
	};

	render() {
		const { menuTitle, menuDescription, children, tile } = this.props;
		return (
			<div styleName={classNames('dropdown-menu', { 'dropdown-menu-tile': tile })}>
				<div
					styleName={`dropdown-menu-modal ${this.handleDropDownClass()}`}
					onClick={this.handleClick}
					role="presentation"
				>
					<div styleName={this.handleLeftContentcontainerClass()}>
						<h2 styleName="dropdown-menu-title"> {menuTitle}</h2>
						<p styleName="dropdown-menu-description"> {menuDescription}</p>
					</div>

					<div styleName="right-content-container">{this.renderDropDownIcon()}</div>
				</div>
				<div styleName="dropdown-menu-content">{children}</div>
			</div>
		);
	}
}

// DropDownMenu.defaultProps = {
// 	leftContentIsSplitted: false,
// 	isContentCompleted: false,
// 	menuTitle: '',
// 	menuDescription: '',
// 	children: null,
// 	isDroppedDown: false,
// 	tile: false,
// 	bindIsDroppedDownStateToParent: value => value
// };

export default CSSModules(DropDownMenu, styles, { allowMultiple: true });
