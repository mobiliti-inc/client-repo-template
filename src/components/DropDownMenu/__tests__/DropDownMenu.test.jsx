import DropDownMenu from '../DropDownMenu';


describe('DropDownMenu', () => {
	it('renders', () => {
		const tree = shallow(<DropDownMenu />);
		matchSnapshot(tree);
	});

	it('renders with splitted left content container', () => {
		const tree = shallow(<DropDownMenu leftContentIsSplitted />);
		matchSnapshot(tree);
	});

	it('updates isDroppedDown state correctly when handleClick method is called', () => {
		const tree = shallow(
			<DropDownMenu />);
		expect(tree.instance().state.isDroppedDown).toEqual(false);
		tree.instance().handleClick();
		expect(tree.instance().state.isDroppedDown).toEqual(true);
	});

	it('updates DropDownClass correctly when isDroppedDown props change', () => {
		const tree = shallow(
			<DropDownMenu />);
		expect(tree.instance().state.isDroppedDown).toEqual(false);
		expect(tree.find('.dropdown-menu-modal-open').length).toEqual(0);
		tree.setProps({
			isDroppedDown: false
		});
		expect(tree.instance().state.isDroppedDown).toEqual(false);
		expect(tree.find('.dropdown-menu-modal-open').length).toEqual(0);
		tree.setProps({
			isDroppedDown: true
		});
		expect(tree.find('.dropdown-menu-modal-open').length).toEqual(1);
	});

	it('updates LeftContentcontainerClass correctly when isContentCompleted props change', () => {
		const tree = shallow(
			<DropDownMenu />);
		expect(tree.instance().state.isContentCompleted).toEqual(false);
		expect(tree.find('.left-content-container-selected').length).toEqual(0);
		tree.setProps({
			isContentCompleted: false
		});
		expect(tree.instance().state.isContentCompleted).toEqual(false);
		expect(tree.find('.left-content-container-selected').length).toEqual(0);
		tree.setProps({
			isContentCompleted: true
		});
		expect(tree.find('.left-content-container-selected').length).toEqual(1);
	});
});
