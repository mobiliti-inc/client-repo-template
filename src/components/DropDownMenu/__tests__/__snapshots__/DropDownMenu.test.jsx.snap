// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownMenu renders 1`] = `
<div
  className="dropdown-menu"
>
  <div
    className="dropdown-menu-modal dropdown-menu-modal-close"
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="left-content-container"
    >
      <h2
        className="dropdown-menu-title"
      >
         
      </h2>
      <p
        className="dropdown-menu-description"
      >
         
      </p>
    </div>
    <div
      className="right-content-container"
    >
      <img
        alt="drop icon"
        src="test-file-stub"
      />
    </div>
  </div>
  <div
    className="dropdown-menu-content"
  />
</div>
`;

exports[`DropDownMenu renders with splitted left content container 1`] = `
<div
  className="dropdown-menu"
>
  <div
    className="dropdown-menu-modal dropdown-menu-modal-close"
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="left-content-container-split"
    >
      <h2
        className="dropdown-menu-title"
      >
         
      </h2>
      <p
        className="dropdown-menu-description"
      >
         
      </p>
    </div>
    <div
      className="right-content-container"
    >
      <img
        alt="drop icon"
        src="test-file-stub"
      />
    </div>
  </div>
  <div
    className="dropdown-menu-content"
  />
</div>
`;
